Add documentation and sample code to use probes

contrun opened the pull request:
This PR is based on https://github.com/nervosnetwork/ckb-standalone-debugger/pull/62. There are two additional probes here. One is used to track syscall parameters and return values. One is used to track function calls and returns. The functionality of the code here has been manually verified with https://github.com/contrun/ckb-vm-bpf-toolkit/blob/731ce2095eb4074e57616dad538ed51bf598f255/tools/uretprobe.py. I used https://github.com/contrun/ckb-standalone-debugger/tree/use-ckb-vm-with-probe-support (which requires https://github.com/contrun/ckb-vm/tree/add-probes-for-syscalls-and-functions that have the same code change to this commit except that is based on v0.22.2) to show the usage of these probes. For example, below is a sample output of running `sudo env RUST_LOG=debug ./tools/memdump.py --bin ../ckb-lua/build/lua-loader.debug -- -e 'print(42)' --bpf-func "^test_func$" --bpf-regs a0 --bpf-debugger ../ckb-standalone-debugger/target/release/ckb-debugger`

```
Func test_func has been jumped to/from 2 times!
Func test_func has been called 1 times!
Func test_func has returned 1 times!
Dumping return value counts for func test_func
return value: 000000000005d020, count: 1
Dumping meomry addresses for func test_func
memory addr: 00007f1474126010, content: 0000000000000001
Dumping meomry content for func test_func
memory addr: 000000000005d020, content: 31302f2e2d2c2b2a
```

contrun committed:
manually execute instruction in probe mode

manually execute instruction in probe mode

Revert "use ckb-vm with probe support"

This reverts commit a47504a0f77cb1411286648290e87134383ec73c.

contrun committed:
add probes for syscalls

contrun committed:
add probe for function calling/return

contrun committed:
add docs on how to use usdt probes

XuJiandong commented on ckb-debugger/docs/probes.md:
````diff
@@ -0,0 +1,100 @@
+## syscalls
+Syscalls are easier to trace, to print the arguments to the syscall, we can just run bpftrace
+```
+sudo bpftrace -e 'usdt:./target/release/ckb-debugger:ckb_vm:syscall /arg0 == 2061/ { printf("%d, %d, %d, %d, %d\n", arg0, arg1, arg2, arg3, arg4); }'
````
Add a link to syscall id: https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0009-vm-syscalls/0009-vm-syscalls.md

XuJiandong commented on ckb-debugger/docs/probes.md:
````diff
@@ -0,0 +1,100 @@
+sudo bpftrace -e 'usdt:./target/release/ckb-debugger:ckb_vm:syscall /arg0 == 2061/ { printf("%d, %d, %d, %d, %d\n", arg0, arg1, arg2, arg3, arg4); }'
+```
+
+Here arg0 is the syscall number, and arg1, arg2, arg3, arg4 are arguments to the syscall. A sample output is
````
Give a mapping between arg0, arg1, ... to syscall argument names.

XuJiandong commented on ckb-debugger/docs/probes.md:
```diff
@@ -0,0 +1,100 @@
+If the destination address of next jal/jalr instruction was found in the hash map, then this instruction is likely a function return. Otherwise, this may be a function call.
+
+The script [trace.py](../examples/trace.py) implements such logic. To trace the function calling and returning, 
+We can run `sudo env RUST_LOG=debug ./example/trace.py --bin ./example-bin --bpf-func "^test_func$" --bpf-debugger ../ckb-standalone-debugger/target/release/ckb-debugger`. Below is a sample output of running this script.
```
This command line is too long. Make it configurable in python or config files. Tell users how to modify according to their needs.

XuJiandong commented on ckb-debugger/docs/probes.md:
```diff
@@ -0,0 +1,100 @@
+Dumping return value counts for func test_func
+return value: 000000000005d020, count: 1
+Dumping meomry addresses for func test_func
+memory addr: 00007f11c6b4b010, content: 0000000000000001
```
A valid CKB address is from 0 ~ 2^32. Why is this one so large?

contrun replied:
The `memory addr` here is the memory address of the host program (ckb-debugger), not the memory address of the program running in ckb-vm. I will delete that to avoid confusion.

xxuejie commented:
Personally, I wasn't sure about adding new probes tracking syscalls/function calls, I'm more opting for another path, where we do opcode parsing within BPF scripts, and deduce if an instruction is a syscall / function call there. A generic probe in the debugger is all we need then. Right now in this PR we are paying for unneeded opcode parsing work even when we are not tracing syscalls. I wasn't sure if this is desired.

contrun commented:
> Personally, I wasn't sure about adding new probes tracking syscalls/function calls, I'm more opting for another path, where we do opcode parsing within BPF scripts, and deduce if an instruction is a syscall / function call there. A generic probe in the debugger is all we need then. Right now in this PR we are paying for unneeded opcode parsing work even when we are not tracing syscalls. I wasn't sure if this is desired.

There are definitely some room for discussion on what should be included in `ckb-debugger`, and I fully agree we should not introduce extra overhead if the tracing feature is not used.

Here is some justification of this PR. For one thing, we can't get syscall return values by just using the `execute_instruction` and `execute_instruction_end` probes, as syscall returns are not really a result of executing an instruction.

For another thing, it is quite hard for typical ckb developers to track function calls and returns with just `execute_instruction` and `execute_instruction_end`. Because firstly, instructions exposed by `execute_instruction` and `execute_instruction_end` are not readily usable, as they are in an internal representation format. We need to decode them first (even decoding the address to be jumped to requires some work). And secondly RISCV jumping/returning mechanism is quite convoluted for everyday ckb contract developers (it is not immediately clear which function jumps/returns to which function based solely on the jal/jalr instructions). I initially tried to make a probe for function calls and another probe for function returns to ease the burden of normal ckb developers. But I found integrating `addr2line` like modules into `ckb-debugger` undesirable. So I decide to make a compromise to add the probe `jump`. There are de

xxuejie commented:
I do get the concerns you are mentioned here, to summarize them a bit, we are on the same page that developers will need some help getting the exact information they need from pure byte streams. The question here, however, is on a different topic: should we build the supporting code in Rust in the debugger, or should we provide the supporting code in BPF code?

An anecdote first: at the beginning, I actually wrote much of the ckb-vm-bpf-toolkit in [bpftrace](https://github.com/iovisor/bpftrace), which uses much lesser lines of code and avoids the mix of C and python in the toolkit right now, but soon I ran into a big hurdle of bpftrace: it does not allow one to define custom operators/functions! To use BPF in ckb-vm, I expect us to develop a non-trivial library supporting all kinds of use, [this file](https://github.com/xxuejie/ckb-vm-bpf-toolkit/blob/7d953ff03509990b99d742b5e1d37bf990fc7776/bpfc/riscv.h) covers only a few macros for now, but it can serve as a foundation, nothing prevents us to add more C functions for different purposes.

With those initial thoughts laid out, let me address each individual point you raise:

* To me, the only value that cannot be captured in `execute_instruction_end` after a syscall completes, is that the value in `a0` will be overriden by the return value of the syscall pseudo function. To overcome this, we can capture the value of `a0` in `execute_instruction` trace via a map variable, then use the captured value in `execute_instruction_end`, where we can safely gather all responded data in a syscall. This is a quite typical convention in BPF scripts. I do want to recommend the famous [bpf book](https://www.brendangregg.com/bpf-performance-tools-book.html) for more patterns like this.
* Decoding instructions certainly requires work of course, but it remains a question if we want to do it in Rust vs. do it in BPF scripts. The added code for decoding instructions in this PR, could well be replicated in C in a BPF script. One can say doing it in the debugger side in Rust is easier and faster because we can pretty much copy the same code from ckb-vm, but it still remains a question if this is an elegant way. To me, those things could well be performed at BPF side, and we should leave the debugger code clean and unchanged.
    + One further point, it is true that the instruction we are exposing here, is previously only internal to ckb-vm. But personally, I see no issues decoding them here, since we are essentially profiling programs running on ckb-vm, and every now and then BPF will definitely require us to know a little internal details of ckb-vm(e.g., `regs` contains a 32-element, 8-byte-each array of all RISC-V registers, can well be viewed as internal details of ckb-vm). Given those reasons, I see no reasons that we shall not decode the *internal* instructions of ckb-vm. But if this really bothers you, nothing prevents us from using `pc` and `memory` to locate the raw RISC-V instruction, and decode it directly.
* Decoding addresses shares a similar story, I do agree we might want to aid the developer somehow, but those logic could well reside in a C function defined in BPF scripts, like in [this file](https://github.com/xxuejie/ckb-vm-bpf-toolkit/blob/7d953ff03509990b99d742b5e1d37bf990fc7776/bpfc/riscv.h). To me we should be working on adding more functions directly here, instead of shortcuting things in Rust at the debugger side.
* As a last point, I strongly vote against introducing `addr2line` or sorts of things in the debugger, I see that as a different path: if we want to build a plugin style API for ckb-debugger, much like the [gdb API](https://sourceware.org/gdb/onlinedocs/gdb/Python-API.html), we might want to add more features in the debugger. However if we are trying out BPF way of doing things, we should stick to BPF philosophy, and build more logics at BPF side. So I introduced `pyelftools` in the `ckb-vm-bpf-toolkit` project.

contrun commented:
I am convinced that we can use bpf for pretty much everything. My another concern is that how can we always distribute all the helper functions and struct definitions to ckb developers. Of course, we can make developers clone ckb-vm-bpf-toolkit and reuse it. I am wondering if there is a simple way to do this, just like extend gdb with python functions.

xxuejie commented:
I don't think distribution at this pint is a problem, we can just keep everything in a [single file](https://github.com/xxuejie/ckb-vm-bpf-toolkit/blob/7d953ff03509990b99d742b5e1d37bf990fc7776/bpfc/riscv.h), due to the way BPF scripts are compiled, we won't run into problems with single header library like typical C projects with multiple `.c` files.

contrun committed:
address the comments in pr review

contrun committed:
decoding ckb instructions in bpf

contrun committed:
decode instructions to find out jump in bpf

contrun committed:
update documentation for probes

contrun committed:
revert all changes in ckb-debugger

contrun commented:
@xxuejie I have removed all the `ckb-debugger` code to trace functions/syscalls returns/calls. I also added a function calls/returns tracing example in the python code.

xxuejie commented on ckb-debugger/examples/trace.py:
```diff
@@ -0,0 +1,224 @@
+from elftools.elf.elffile import ELFFile
+
+# Change the path of ckb-debugger, the path of the program to run and the function name to probe here.
+BIN_TO_TRACE = "/home/e/Workspace/ckb-lua/build/lua-loader.debug"
```
Please get rid of those hardcoded path, if we want to make this a usable example for everyone, it might be better to introduce command line arguments for providing binary path as well as debugger path.

contrun replied:
I felt that my current example relies too much on the program to be traced. For example, I assumed that the return value of the function to be traced is a memory address and I saved the content to a bpf table. Unlike other examples you provided in https://github.com/xxuejie/ckb-vm-bpf-toolkit, the user of this example has to modify the code anyway.

xxuejie replied:
In that sense I don't think we should put the example code here, it would be better as the accompanying code for a tutorial on running bpf scripts

contrun committed:
move code location
